src/style/style_layer.js

```diff
@@ -162,6 +186,7 @@ class StyleLayer extends Evented {
     }
 
     recalculate(parameters: EvaluationParameters) {
+        this._parameters = parameters;
```

Comment:
we can get rid of it if we pass the evaluation parameters from `Map#_render` to `painter#render` and through to the draw calls. Is that preferable?

Reply:
I took a deeper look and understand things a lot better now: `CrossFadedDataDrivenProperty` wants access to the same state -- `fromScale`, `toScale`, and `t` -- that `CrossFadedProperty` returns via `CrossFaded`. So here's what you could do support them both in a common way:

* Rename `this._parameters` to `this._crossfadeParameters` and change the type to `CrossfadeParameters`.
* Move `getCrossfadeParameters()` to `EvaluationParameters`, so the above line in the PR becomes `this._crossfadeParameters = parameters.getCrossfadeParameters()`. (This isn't strictly necessary, but `getCrossfadeParameters()` seems like it should live on `EvaluationParameters` per [Feature Envy](https://refactoring.guru/smells/feature-envy).)
* Remove `fromScale`, `toScale`, and `t` from `CrossFaded<T>` and the duplicate calculation of these values from `CrossFadedProperty<T>::_calculate`. Instead, have layers that use either `CrossFadedProperty` or `CrossFadedDataDrivenProperty` obtain the calculated values from the same source: `this._crossfadeParameters`.

Reply:
Seems like `CrossFaded<T>` should become `{min: T, mid: T, max: T}`. This would allow it to be used in more places, and maybe DRY up some of the duplication in the bucket classes.

Reply:
@jfirebaugh `{min, mid, max}` are only needed for data-driven property evaluation, and `to`, `from` is still useful for constant and camera expressions, and since the expression evaluation in the binders returns `T`, not `CrossFaded<T>` I've kept the `to`, `from` values for now. 

I DRYd up the bucket code in https://github.com/mapbox/mapbox-gl-js/pull/6289/commits/f1a06a333e061634ba658251fb5cdf277146f10b and took your suggestions for reworking `getCrossfadeParameters` in https://github.com/mapbox/mapbox-gl-js/pull/6289/commits/9f6bac9453f66c163d43b3dc824a18fa33228826. Thank you!!